go-controller/pkg/ovn/controller/podset/controller.go

```diff
@@ -0,0 +1,423 @@
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+)
+
+const controllerName = "ovn-pod-set"
+const maxRetries = 15
+
+// PodSetController is a controller that maps selectors of pods to address sets.
+// The primary use case is for network policies - NetworkPolicy pods and the peers
+// to which it applies are implemented as AddressSets in ovn. This is controller
+// is responsibile for ensuring the address sets in OVN match their selector.
+//
+// On startup, we'll re-build each set and apply it directly to the database;
+// subsequent updates will be processed as deltas.
+type PodSetController struct {
+	podInformer coreinformers.PodInformer
+	podLister   corelisters.PodLister
+	podsSynced  cache.InformerSynced
+
+	namespaceInformer coreinformers.NamespaceInformer
+	namespaceLister   corelisters.NamespaceLister
+	namespacesSynced  cache.InformerSynced
+
+	queue workqueue.RateLimitingInterface
+
+	podSetsLock sync.RWMutex
+
+	// all podSets we know about
+	podSets map[string]*PodSet
+
+	// any podSets that are added before our informers are synced
+	podSetsToSync map[string]interface{}
+}
+
+func NewController(
+	podInformer coreinformers.PodInformer,
+	namespaceInformer coreinformers.NamespaceInformer,
+) *PodSetController {
+
+	psc := &PodSetController{
+		podInformer: podInformer,
+		podLister:   podInformer.Lister(),
+		podsSynced:  podInformer.Informer().HasSynced,
+
+		namespaceInformer: namespaceInformer,
+		namespaceLister:   namespaceInformer.Lister(),
+		namespacesSynced:  namespaceInformer.Informer().HasSynced,
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerName),
+
+		podSets:       map[string]*PodSet{},
+		podSetsToSync: map[string]interface{}{},
+	}
+	return psc
+}
+
+// Run starts the PodSetController, which syncs the caches and applies any
+// PodSets that were added during that time.
+func (psc *PodSetController) Run(threadiness int, stopCh <-chan struct{}) error {
+	defer utilruntime.HandleCrash()
+	defer psc.queue.ShutDown()
+
+	klog.Infof("Starting controller %s", controllerName)
+	defer klog.Infof("Shutting down controller %s", controllerName)
+
+	psc.podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    psc.onPodAdd,
+		UpdateFunc: psc.onPodUpdate,
+		DeleteFunc: psc.onPodDelete,
+	})
+	psc.namespaceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    psc.onNamespaceAdd,
+		UpdateFunc: psc.onNamespaceUpdate,
+		DeleteFunc: psc.onNamespaceDelete,
+	})
+
+	// Wait for the caches to be synced
+	klog.Info("Waiting for informer caches to sync")
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, psc.podsSynced, psc.namespacesSynced) {
+		return fmt.Errorf("error syncing cache")
+	}
+
+	// sync any PodSets that arrived while we were waiting for informer sync
+	go psc.syncPendingPodSets(stopCh)
```

Comment:
This shouldn't be required. Once the cache is in sync, your workqueue will be populated and you can start processing.

Reply:
If anyone called `AddPodSet` while we were still syncing, then we want to make sure we sync that.

We can't apply pod sets until after the informers are sunk, because we need to be able to determine all pods that belong in the set; for that we need `List()` to be complete.

In practice this is academic, since I believe we wait for the informers to sync before we go ahead and start this controller anyways.

Reply:
> In practice this is academic, since I believe we wait for the informers to sync before we go ahead and start this controller anyways.

Yeah. I suppose you could also make AddPodSet return an error if the informers haven't synced yet just in case.